presto-main/src/test/java/com/facebook/presto/execution/TestNodeScheduler.java

```diff
@@ -424,6 +424,47 @@ public void testSplitCount()
         assertEquals(nodeTaskMap.getPartitionedSplitsOnNode(chosenNode), 0);
     }
 
+    @Test
+    public void testMaxTasksPerStage()
+    {
+        NodeTaskMap nodeTaskMap = new NodeTaskMap(finalizerService);
+        TestingTransactionHandle transactionHandle = TestingTransactionHandle.create();
+        NodeSchedulerConfig nodeSchedulerConfig = new NodeSchedulerConfig()
+                .setMaxSplitsPerNode(20)
+                .setIncludeCoordinator(false)
+                .setMaxPendingSplitsPerTask(10);
+
+        NodeScheduler nodeScheduler = new NodeScheduler(new LegacyNetworkTopology(), nodeManager, nodeSchedulerConfig, nodeTaskMap);
+        NodeSelector nodeSelector = nodeScheduler.createNodeSelector(CONNECTOR_ID, 2);
+
+        ImmutableList.Builder<Split> initialSplits = ImmutableList.builder();
+        for (int i = 0; i < 10; i++) {
+            initialSplits.add(new Split(CONNECTOR_ID, transactionHandle, new TestSplitRemote()));
+        }
+
+        Set<Split> splits = new HashSet<>();
+        for (int i = 0; i < 5; i++) {
+            splits.add(new Split(CONNECTOR_ID, transactionHandle, new TestSplitRemote()));
+        }
+        Map<InternalNode, RemoteTask> taskMap = new HashMap<>();
+        Multimap<InternalNode, Split> assignments = nodeSelector.computeAssignments(splits, ImmutableList.copyOf(taskMap.values())).getAssignments();
+        MockRemoteTaskFactory remoteTaskFactory = new MockRemoteTaskFactory(remoteTaskExecutor, remoteTaskScheduledExecutor);
+        int task = 0;
+        for (InternalNode node : assignments.keySet()) {
+            TaskId taskId = new TaskId("test", 1, 1, task);
+            task++;
+            MockRemoteTaskFactory.MockRemoteTask remoteTask = remoteTaskFactory.createTableScanTask(taskId, node, ImmutableList.copyOf(assignments.get(node)), nodeTaskMap.createPartitionedSplitCountTracker(node, taskId));
+            remoteTask.startSplits(25);
+            nodeTaskMap.addTask(node, remoteTask);
+            taskMap.put(node, remoteTask);
+        }
+        assertEquals(assignments.keySet().size(), 2);
+        Multimap<InternalNode, Split> tmp = assignments;
+        assignments = nodeSelector.computeAssignments(splits, ImmutableList.copyOf(taskMap.values())).getAssignments();
+        // second call should return same nodes
+        assertEquals(assignments.keySet(), tmp.keySet());
```

Comment:
I wonder if it makes sense to exercise more corner cases here:

*
1. Select one node
2. Call the second time, make sure than the second node is selected, and it is different than the first one
3. Call the third time, make sure that no additional nodes have been selected.

*
1. Add only a single node to the InmemoryNodeManager
2. Call once, make sure it is selected
3. Call the second time, make sure the output is the same.
4. Add one more node to the InmemoryNodeManager
5. Call the third time, make sure it has a new node selected
6. Add one more node to the InmemoryNodeManager
7. Call for the forth time, make sure the latest added node is not selected, as the limit is reached.

Reply:
@arhimondr I couldnt understand the first test case. select one node twice?

Reply:
e.g.: provide only a single split. So only a single node of three is selected at the first step.